File: problems/0076-minimum-window-substring/solution.py
# Link: https://leetcode.cn/problems/minimum-window-substring/
# Level: Hard
# Date: 2025-04-16

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        t_cnt = defaultdict(int)
        for c in t:
            t_cnt[c] += 1
        
        required = len(t_cnt)
        formed = 0
        s_cnt = defaultdict(int)
        left = 0
        min_len = float('inf')
        ans_left = 0

        for right in range(len(s)):
            c = s[right]
            s_cnt[c] += 1

            if c in t_cnt and s_cnt[c] == t_cnt[c]:
                formed += 1

                # 当窗口满足条件时，尝试收缩左边界
                while left <= right and formed == required:
                    if right - left + 1 < min_len:
                        min_len = right - left + 1
                        ans_left = left
                    
                    left_c = s[left]
                    s_cnt[left_c] -= 1
                    if left_c in t_cnt and s_cnt[left_c] < t_cnt[left_c]:
                        formed -= 1
                    left += 1
        
        return s[ans_left:ans_left+min_len] if min_len != float('inf') else ""

File: problems/0076-minimum-window-substring/solution.cpp
// Link: https://leetcode.cn/problems/minimum-window-substring/
// Level: Hard
// Date: 2025-04-16

class Solution {
    private:
        bool check(unordered_map<char, int>& t_cnt, unordered_map<char, int>& s_cnt) {
            for (auto it = t_cnt.begin(); it != t_cnt.end(); it++) {
                if (s_cnt[it->first] < it->second) {
                    return false;
                }
            }
            return true;
        }

    public:
        string minWindow(string s, string t) {
            if (s.size() < t.size()) {
                return "";
            }
            unordered_map<char, int> t_cnt;
            for (char c : t) {
                t_cnt[c]++;
            }
            int required = t_cnt.size();
            unordered_map<char, int> s_cnt;
            int left = 0, right = 0;
            int min_len = INT_MAX;
            int ans_left = 0;
            int formed = 0;
            while (right < s.size()) {
                char c = s[right];
                s_cnt[c]++;
                if (t_cnt.count(c) && s_cnt[c] == t_cnt[c]) {
                    formed++;
                }

                while (formed == required && left <= right) {
                    if (right - left + 1 < min_len) {
                        min_len = right - left + 1;
                        ans_left = left;
                    }

                    char left_c = s[left];
                    s_cnt[left_c]--;
                    if (t_cnt.count(left_c) && s_cnt[left_c] < t_cnt[left_c]) {
                        formed--;
                    }
                    left++;
                }
                right++;
            }

            return min_len == INT_MAX ? "" : s.substr(ans_left, min_len);
        }
};
